Add checking-bitcoin-balances-privately review to wasabi-research-club (#643)
bitcointranscripts/bitcointranscripts: wasabi-research-club/checking-bitcoin-balances-privately.md

Reviewer: `Chaumian`. Not ~~Chowmian~~

Reviewer: ```suggestion
Have you looked into the open source hardware's Verilog synthesis stuff like IceStorm or something?
```

Reviewer: ```suggestion
Blind signatures probably did a p-set about that, but I could not off the top of my head say what exactly it is.
```

slang for "problem set"

Reviewer: ```suggestion
Like that floating point error where you add 1.00001 and 1.00001 and you get 2, because of that floating point error, it's analogous to that.
```

Reviewer: floating pointer => floating `point error`

Author: what's this comment for ?

Reviewer: `floating "pointer"` is not correct. I checked and listened to the YouTube segment. Please Search the phrase in the transcript. GitHub suggested change doesn't work properly.

<img width="502" height="63" alt="image" src="https://github.com/user-attachments/assets/e9084cf1-712b-48bf-9367-ed590f07ce8f" />

---
title: 'Checking Bitcoin balances privately'
transcript_by: 'Garvit-77 via review.btctranscripts.com'
media: 'https://www.youtube.com/watch?v=8L725ufc-58'
date: '2022-09-27'
tags:
  - 'research'
  - 'privacy-enhancements'
  - 'cryptography'
speakers:
  - 'Samir Menon'
  - 'Lucas Ontivero'

>   - 'Max Hillebrand'

Reviewer: Add Adam Ficsor

Reviewer: Add Jumar Macato ?

Jumar is a visual graphics design programmer working on the Avalonai UI Framework, he is a lead developer and also developer at Wasabi Wallet working on the User Interface of the Wallet and its UX aspects of it. 

https://podcasts.apple.com/us/podcast/jumar-macato-from-humble-roots-to-avalonia-and/id1552688883?i=1000582084418

summary: 'In this video, the Wasabi Research Club delves into the topic of checking Bitcoin balances privately. They discuss the use of homomorphic encryption and private information retrieval to protect users'' privacy when querying data from a server. They explore different solutions and strategies to address challenges such as address linkability, block retrieval, and the scalability and cost of checking balances. The team also emphasizes the need for an open standard for private information retrieval to ensure decentralization. They discuss the use of Tor and homomorphic encryption together, as well as the possibility of batching requests to improve efficiency. The video concludes by highlighting the ongoing research and exploration of potential solutions to ensure private querying of Bitcoin balances and more...'
aliases:
  - '/wasabi/research-club/checking-bitcoin-balances-privately/'
---
**Max Hillebrand:** 00:00:00

So hello and welcome to the Wasabi Wallet Research Club.
Today we are speaking with Samir from Spiral, which is the title of a fancy cryptography paper of homomorphic value encryption or homomorphic encryption and private information retrieval.

> The gist of this `cryptomagic` is that a client can request data from a server, but the server does not know which data was requested.

Reviewer: ```suggestion
The gist of this cryptomagic is that a client can request data from a server, but the server does not know which data was requested.
```

> And there are different variants of the `cryptomagic` for different use cases.

Reviewer: ```suggestion
And there are different variants of the cryptomagic for different use cases.
```

Reviewer: The backticks (``) should be removed. It is not Bitcoin or cryptography terminology. It is simply a playful phrase used by the host

And the second, more pressing and important for us, is an anonymous block explorer so that you can query the UTXO set and request an address and the server will give you the balance of that address.
And well, the server doesn't know which address you're actually requesting.
So that sounds like impossible magic.
And Samir, please tell us how all the magic works.

**Samir Menon:** 00:01:26

Yeah, that was a great introduction.
That's kind of what got me into this field is that it does sound like magic and it does sound like you probably shouldn't be able to do something like that, but you can.
I'll talk a little bit about the internals of how Spiral works at the end, but I'll kind of first just like, let's just try to situate, the security model of a server that doesn't learn what you query is kind of complicated and so let's just first try to compare it to all the other ways you can check a Bitcoin balance.
And let's just like situate it amongst the many, many alternative ways, one could look up the balance of a Bitcoin address.

> A little background on me, I was a graduate student at Stanford and this PAR project was kind of my senior project for my master's and it was advised by Dan Bonet but I went and got a real job, and so I stopped working on it for a little bit.

Reviewer: PIR project? (Private Information Retrieval)
Dan Boneh?
https://en.wikipedia.org/wiki/Dan_Boneh

> A little background on me, I was a graduate student at Stanford and this PIR project was kind of my senior project for my master's and it was advised by Dan Bonet but I went and got a real job, and so I stopped working on it for a little bit.

Reviewer: ```suggestion
A little background on me, I was a graduate student at Stanford and this PIR project was kind of my senior project for my master's and it was advised by Dan Boneh but I went and got a real job, and so I stopped working on it for a little bit.
```

Reviewer: https://en.wikipedia.org/wiki/Dan_Boneh

And yeah, of course, the code is open source, and there's a paper and everything.
And I'm happy to answer any questions about how it works.
I have a slide that kind of explains more of the technical underpinnings of `Spiral`.
So if you want, I can just go through that slide if you want to hear more about `homomorphic encryption` stuff.

## Open Questions

**Samir Menon:** 00:12:00

But before I do that, I'll just kind of say, yeah, I think the open questions for us are what minimum set of data is enough?
I think balance is on its own, it's not quite enough.
I think if I was, When I run a wallet software, I kind of expect more than just the balance in all my addresses.
We obviously need to think about fetching more than one address.
Because today there's like a website and you can fetch one address, but wallets have more than one address.
And yeah, I think there's a couple options on how we do that.
We need to think a little bit about the pay or incentivization structure for servers.
This is a kind of class in computation.
So how can we make it feasible or practical for them?
And long term, I think I would like to see an open standard for PIR for this data.

> So what we want is something that's not tied to a company or a person or an organization, but just kind of ideally, like maybe a BIP or something that allows us to not tie ourselves to any particular scheme and do this PAR thing as an extension of our current way doing a get block RPC or whatever.

Reviewer: ```suggestion
So what we want is something that's not tied to a company or a person or an organization, but just kind of ideally, like maybe a BIP or something that allows us to not tie ourselves to any particular scheme and do this PIR thing as an extension of our current way doing a get block RPC or whatever.
```

PIR? (Private Information Retrieval)

So you wouldn't say that.

**Max Hillebrand:** 00:18:28

Just as a heads up for you, we're downloading the filters from our server, like the Wasabi backend server, but then the blocks are downloaded from the Bitcoin peer-to-peer network.
So the server doesn't incur the block download cost.

**Samir Menon:** 00:18:41

Oh, okay.
So is the main cost for you guys right now actually the filters because you incur that cost as I see.
Yeah, it's a large outgoing cost, I'm sure your hosting provider is charging you.
Actually doing proofs of inclusion is possible, it's good to know you need that kind of to deploy this for real.
If I can ask a follow-up question, do you guys, to do this, are you mostly querying the UTXO set, the full set of transactions, just balances?
Like what kind of data is crucial?

**Max Hillebrand:** 00:19:30

We do want the full transaction history list and so that's what we get in the filters right now.
Wasabi is `SegWit` only, so we don't have to create filters `pre-SegWit`, August 2018 or something, or 17.

**Audience 1:** 00:19:57

The filters are all essentially a compact representation of all `BECH-32` addresses in a block?

> Single public key `BECH-32` addresses.

Reviewer: `Bech32` ?

https://learnmeabitcoin.com/technical/keys/bech32/
https://bitcoinops.org/en/topics/bech32/

Hello, guys.
It's quite nice to be here.
I don't have any strong background in crypto and higher order math, but as a layman, is there a way to simplify the explanation of how homomorphic encryption works?
I have tried to see the information on the internet, but it's just not possible for a basic layman like me.
Would that be possible on this call, or is it not?

**Samir Menon:** 00:22:17

Yeah, I have a slide.
Let's talk a little bit more about Wasabi Wallet and privacy, but then I can give that explanation.
Does that sound good?

**Audience 2:** 00:22:28

Yes, thank you very much.
Thank you.

**Max Hillebrand:** 00:22:34

What I do wonder is, you do need to know the input, like the value of the UTXO that you're trying to spend.
Does your database include amounts?

**Samir Menon:** 00:22:49

Yes.
So the total size of our database is roughly one gig right now.
We translate every, all UTXOs kind of reduced down to about a gig of data.
If we instead made the set all UTXOs and we didn't just take the five most recent ones for every address, the database would not be much larger.
It would be like a small multiple, like maybe three or four times larger, that would be kind of feasible.

> I guess when I'm thinking about it, I suppose it helps that you guys only support the `BECH-32` addresses, because That means there's kind of a limited set of addresses that are on chain that have transactions for that.

Reviewer: `Bech32` ?

But yeah, it's around 300 megabytes per second.

**Audience 1:** 00:15:15

I see.
So the big situation here is that for WasabiWallet, we would want, as like an MVP for us to work, the amounts simply wouldn't be enough.
Our users, they want proof.
They want to know that it's connected to block headers that are in the chain with the most proof of work.
But on top of that, we also need to know that the server isn't in some ways deceiving us.
For example, just lowering the balances of all the users or just omitting certain UTXOs. Block filters do a nice job of this because you're, you know, you're just hoping the server does an accurate job of creating the filters and you're clearing the blocks and you get these entire blocks.
So the only way that you're going to have an incorrect balance is if the server somehow malleates a filter but there's not really good, it's kind of a weird, unclear attack vector.
So how practical would it be to actually get some kind of proofs on top of the balance that you're already producing?

**Samir Menon:** 00:16:35

Yeah, that's a great question.
So, I'm sure the exact name of this is so, so what we want is a `Merkle inclusion proof`, right?

> ## Miracle Inclusion Proof

Reviewer: ```suggestion
## Merkle Inclusion Proof
```

**Samir Menon:** 00:28:35

I see.
So are the addresses that you create there, are they unspent?
I mean, If you create an address and no one hears about it, did it really get created?

**Max Hillebrand:** 00:28:50

Well, in this case, yes, because the `CoinJoin` coordinator hears about it.
And probably also the other `CoinJoin` signers.
So it's semi-public.

**Samir Menon:** 00:29:00

Will it store value?
It will, right?
I guess to start the `CoinJoin` it has some value in it.

**Max Hillebrand:** 00:29:09

No, sorry.
So we spend inputs that are addresses with money on them, but then on the output side we create new addresses that are not yet used without money on it.

**Samir Menon:** 00:29:18

Oh, I see.
So they can be empty output addresses.

**Max Hillebrand:** 00:29:22

Exactly.

> Those are addresses that never were on the blockchain in an output with any amount of stats.

Reviewer: ```suggestion
Those are addresses that never were on the blockchain in an output with any amount of sats.
```

And you might have noticed it's pretty big.
It's like eight megabytes or something.

**Max Hillebrand:** 00:33:59

And Does that size depend on the database size?

**Samir Menon:** 00:34:02

It does not or it only logarithmically does.
So it's like very, I mean, like if the database was 100 times bigger, it would be like 12 megabytes or something.

**Max Hillebrand:** 00:34:17

> Yeah, like, I mean, just general, like, it's a very broad question, but is it, so if we want to have the full TX outset of all, like, of, but actually, we probably also want transaction IDs and stuff like this so basically we want the full transaction metadata blockchain thing for all `SegWit` and `Taproot` outputs yep and let's say we have I don't know 10,000 users or so and each of them has let's say, a thousand addresses or so.

Reviewer: ```suggestion
Yeah, like, I mean, just general, like, it's a very broad question, but is it, so if we want to have the full TXO set of all, like, of, but actually, we probably also want transaction IDs and stuff like this so basically we want the full transaction metadata blockchain thing for all `SegWit` and `Taproot` outputs yep and let's say we have I don't know 10,000 users or so and each of them has let's say, a thousand addresses or so.
```

TXO set (Transaction Output set)
https://gnusha.org/pi/bitcoindev/20170224010943.GA29218@savin.petertodd.org/

> Yeah, like, I mean, just general, like, it's a very broad question, but is it, so if we want to have the full TXO outset of all, like, of, but actually, we probably also want transaction IDs and stuff like this so basically we want the full transaction metadata blockchain thing for all `SegWit` and `Taproot` outputs yep and let's say we have I don't know 10,000 users or so and each of them has let's say, a thousand addresses or so.

Reviewer: ```suggestion
Yeah, like, I mean, just general, like, it's a very broad question, but is it, so if we want to have the full TXO set of all, like, of, but actually, we probably also want transaction IDs and stuff like this so basically we want the full transaction metadata blockchain thing for all `SegWit` and `Taproot` outputs yep and let's say we have I don't know 10,000 users or so and each of them has let's say, a thousand addresses or so.
```

Reviewer: "full TXO set" refers to the complete Transaction Output set

So you, sorry, the input amount of that transaction, because that's on the blockchain, not in the transaction itself, right?
It's on the previous transaction output.
And then you cannot do effective fee bumping in a child pays for parent transaction.
So this is an issue that we have of kind of quote-unquote stuck payments that you received.
And maybe something like that might be helpful.
And so you can query the amount of that input and then do better fee bumping.

**Samir Menon:** 00:39:24

So to summarize, You know you received a payment, but you just don't know how much it is, essentially, because that information is just kind of not local.

**Max Hillebrand:** 00:39:38

No, so you receive a transaction where on the output side your address is.
That's why the transaction was hit in the filter.
But you don't know the value of the inputs exactly, because that's in the previous transaction output.
So you don't know the fee rate that this current transaction has.

**Samir Menon:** 00:40:00

You know that it's less than or equal to, but you don't know what gap was left for the minor in the form of the fee.

**Max Hillebrand:** 00:40:08

> Exactly and then you can't do child pays for parent fee calculations.

Reviewer: ```suggestion
Exactly and then you can do child pays for parent fee calculations.
```
Sounds like "can"?

https://bitcoinops.org/en/topics/cpfp/

**Samir Menon:** 00:51:52

Yeah, it's a great question.
So no.
So all the orange blocks have the exact same size.
So that's a very good point.
Like this vector of encrypted 0's and 1's, it's not small, Because every one of them is big enough to hold a plain text element, so it's pretty big.
There are a bunch of things you can do to kind of not actually send encrypted zeros.
So I guess I can just explain it this way.
So the operations we have are homomorphic multiplication by a plaintext item and addition.
I don't know your familiarity with linear algebra or something, but that's a not quite complete set.
If you were also able to multiply two encrypted numbers.
You can think about it and see that actually you can compute any function on encrypted data if you have addition and multiplication.
Every computation is like a composition of adding and multiplying.
All our computers do are add and multiply.
So there are ways to compute an arbitrary function on encrypted data.
So what that means is the query is now much smaller and you perform this arbitrary computation to make all of these encrypted zeros and ones in a big way while keeping the query small.

**Max Hillebrand:** 00:53:39

> Let's go to more.

Reviewer: ```suggestion
What's up, Jumar?
```

Reviewer: <img width="1610" height="676" alt="Image" src="https://github.com/user-attachments/assets/07f9bf34-30d6-4e60-8e75-e064ae61d7a2" />

<img width="1623" height="456" alt="Image" src="https://github.com/user-attachments/assets/440401d0-6239-4472-a8ee-4fd46bfcd61a" />

> **Audience 2:** 00:53:42

Reviewer: Jumar Macato

OK, this is where I saw this illustration, because I sent maybe the link to it before.
My question is what is the mathematical principle that enables this?
I'm a bit familiar with linear algebra.

**Samir Menon:** 00:54:09

So the main principle is that it's difficult to kind of invert matrices when they have noise.
So I can say more about that, but basically what that means is like if you have a matrix, a large random matrix, and you multiply it by a vector, The outputs are easy to determine.
The output and the input vector are kind of correlated.
It's easy to invert the matrix and figure out what the input was from the output.
So if I give you A, S and I give you A, it's easy to find S, because finding the inverse of a matrix is not hard.
But the key insight, the algebraic insight that enables this is that, Well, if I give you A S plus E, where E is just some very small noise, suddenly the problem is very hard.
It's very hard to figure out S from A S plus E.
The post has further down kind of a discussion of like learning with errors and stuff.

> **Audience 2:** 00:55:23

Reviewer: Jumar Macato

You can't do it an unlimited number of times.
Otherwise, the ciphertext will kind of become garbage.

> **Audience 2:** 00:57:19

Reviewer: Jumar Macato

Okay, that's the way you explained it in the terms of matrices.
Yeah, I think it clicks something on me.

**Samir Menon:** 00:57:28

Yeah, feel free to reach out if you have other questions.

> **Audience 2:** 00:57:33

Reviewer: Jumar Macato

So basically what we do is we bound the standard deviation of the output noise.
And by doing that, we can say the noise will always be within six standard deviations of the mean and we know the standard deviation is going to be x and so we know that we will decrypt correctly.

**Audience 2:** 00:59:11

Okay, cool.
That's all for me.
Thank you so much.

> Speaker 4: 00:59:19

Reviewer 2: There should be a speaker name

Author: Thanks for Helping me there

So basically what we do is we bound the standard deviation of the output noise.
And by doing that, we can say the noise will always be within six standard deviations of the mean and we know the standard deviation is going to be x and so we know that we will decrypt correctly.

> **Audience 2:** 00:59:11

Reviewer: Jumar Macato ?

So yeah.
But obviously for an open standard, we're going to need open source clients and servers.

**Max Hillebrand:** 01:00:35

Cool, thank you.

> **Audience 5:** 01:00:36

Reviewer: ```suggestion
**Adam Ficsor:** 01:00:36
```

Reviewer: https://github.com/bitcointranscripts/bitcointranscripts/pull/643/changes#r3249755893

"nopara73" is the well-known online pseudonym for Wasabi Wallet founder Adam Ficsor

So I just arrived.

> Is there a way to summarize whole Spiral work in a, like, like on five?

Reviewer: ```suggestion
Is there a way to summarize whole Spiral work in a, like, like I'm five?
```

Reviewer: <img width="1970" height="1118" alt="Image" src="https://github.com/user-attachments/assets/31dab4ce-f9a1-40d4-b95e-8ea431a44561" />

So I just arrived.
Is there a way to summarize whole Spiral work in a, like, like on five?

**Samir Menon:** 01:00:54

That's magic.
Yeah, I think the summary, the best summary is honestly Max's at the beginning, which was just like, basically Spiral is a way to retrieve an item from a database without letting it ever learn what item you retrieved.

> Speaker 4: 01:01:14

Reviewer: Speaker 4 ?

> Speaker 4: 01:02:12

Reviewer: ?

Author: I'm confused what this comment is about

I think it gets maybe really bad if you are trying to query tens of thousands of addresses.
Then we'll have to think of a better way to kind of batch your queries.

**Max Hillebrand:** 01:04:38

Yeah, so Adam, there's multiple layers of why this might be very difficult for us to use.
One is we might want the transaction history, so this is the whole TX outset instead of the UTXO set.

> Speaker 4: 01:04:49

Reviewer: ?

We definitely want that, so we cannot get that?

**Max Hillebrand:** 01:04:55

We could, but then the size of the database gets larger because you need to store all outputs, not just the unspent ones.
And so that's the first issue.

> Speaker 4: 01:05:05

Reviewer: ?

I think it gets maybe really bad if you are trying to query tens of thousands of addresses.
Then we'll have to think of a better way to kind of batch your queries.

**Max Hillebrand:** 01:04:38

Yeah, so Adam, there's multiple layers of why this might be very difficult for us to use.

> One is we might want the transaction history, so this is the whole TX outset instead of the UTXO set.

Reviewer: ```suggestion
One is we might want the transaction history, so this is the whole TXO set instead of the UTXO set.
```

You could just query that database and you could find out, okay, 9,000 of these addresses, nothing happened on them since I last opened this app.
And then for the thousand that are remaining, you could look and see, which, which ones of them have a balance still or are still active.
And then you could retrieve each of their transaction histories kind of individually.
It's not perfect, but something I'll do is I think I will l try to build up a number of addresses in my wallet that's large enough to kind of test with.
And I would love to play around and see how this could work or at least, yeah.
Yeah, I think it's a kind of urgent challenge.
So, yeah, and it's cool tech.

## Private Information Retrieval from a Database

**Max Hillebrand:** 01:08:20

Another interesting, like, the cool thing is like private information retrieval from a database.
It seems massively big as a concept.
And we can use it in many different areas.
One, for example, might be to query the mempool of someone else, which you boot up your client and you have unconfirmed transactions and this way you could be able to get them.

**Adam Ficsor:** 01:08:45

> So I did that in hidden wallet.

Reviewer: ```suggestion
So I did that in HiddenWallet.
```

Reviewer: https://www.ccn.com/bitcoin-privacy-takes-another-step-forward-with-wasabi-wallet-launch/

> Ficsor unveiled [Wasabi](https://medium.com/@nopara73/wasabi-privacy-focused-bitcoin-wallet-for-desktop-3962d567045a) at Building on Bitcoin 2018, an industry conference dedicated to technical developments, explaining that it is a new implementation of the project formerly known as HiddenWallet.

> So I did that in hiddenWallet.

Reviewer: ```suggestion
So I did that in HiddenWallet.
```

Reviewer: Do you know how to use applying suggested change?
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

From the previous review:
https://github.com/bitcointranscripts/bitcointranscripts/pull/643#discussion_r3255241091

> Speaker 4: 01:08:45

Reviewer: ?

> Speaker 4: 01:09:35

Reviewer: ?

It's not gonna be like that well-known fact that they are going for block explorers to disclose data.
You know?

**Samir Menon:** 01:11:00

Yeah, I mean, if I was, I would definitely subpoena Block Explorer.
I think Ledger keeps their logs on Ledger Live for like five years or something.
So yeah, I don't know if you guys are interested.
I think that's another angle we're interested in is this hardware wallet kind of integration.
I think it's kind of funny how you go to kind of you're currently able to go to Greatlands and kind of keep all your private keys on hardware.
But then when you want to just like know how much money you have, you have to just, you end up kind of just going to a block explorer or ledger or whoever, and kind of just telling them your address and your IP address and stuff.

> **Audience 4:** 01:11:46

Reviewer: ```suggestion
**Peter:** 01:11:46
```

Reviewer: I have no idea who this Peter was.

<img width="2234" height="727" alt="Image" src="https://github.com/user-attachments/assets/1a7919e9-62fe-4f75-aaab-a2f18c505d91" />

You could connect it to your own node, right?
I guess, to it and stuff, so yeah.

**Samir Menon:** 01:11:53

Especially from mobile clients and stuff.
Yeah, no, go ahead.

**Max Hillebrand:** 01:11:58

You know, even when you connect to your own node, which is the scenario you brought up earlier.
But the problem with this is you still put your sensitive information on another computer.
And sure, you're the one mainly in control of that computer, but maybe you get hacked, or maybe you are already hacked.
And So the less information you store on someone or on another computer, the better.

> Like an existing example is ElectrumX versus ElectrumPersonalServer.

Reviewer: ```suggestion
Like an existing example is ElectrumX versus Electrum Personal Server.
```

Reviewer: https://docs.btcpayserver.org/ElectrumPersonalServer/

> [Electrum Personal Server or EPS (opens new window)](https://github.com/chris-belcher/electrum-personal-server)is a personal version of public Electrum servers like [ElectrumX](https://docs.btcpayserver.org/ElectrumX/).

> ElectrumPersonalServer stores your extended public key on the server versus ElectrumX only receives an address and then forgets it later.

Reviewer: ```suggestion
Electrum Personal Server stores your extended public key on the server versus ElectrumX only receives an address and then forgets it later.
```

It's going to be tough to, to want to pay a dollar every time you want to know how much money you have, you know?
So, I believe we can work on this.
I'm not, I haven't lost all hope.
I think it's really useful to hear that people have so many addresses and I want to know more about like what those addresses look like.
You kind of explained that they're kind of the outputs of a, their output addresses as a result of a CoinJoin, sometimes a failed CoinJoin.
But yeah, the more I can kind of understand that, the more I think it could be feasible kind of to make progress.

**Max Hillebrand:** 01:17:12

Yeah, but I would guess the vast majority of these addresses are unused.
So they never made it on the blockchain.
And then only a rather smaller number made it on the blockchain and is already spent.
And so not in the UTXO set anymore.
And then the smallest percentage is going to be the unspent UTXO addresses.

**Samir Menon:** 01:17:38

Okay.
Certainly, I mean, spent.
So something cool about spent coins is that once they're spent, they can't come back.
So yeah, we can exploit that.

> Speaker 4: 01:17:56

Reviewer: ?

Lucas, what's your take about all this?
Can we use Spiral somehow we use it to make Wasabi like really light?
Or even should we, even if we could work it out?

> Mike, you are the one who wrote the Columbus Rites Theater stuff, so you put a lot of thoughts into around here.

Reviewer: https://www.youtube.com/watch?v=8L725ufc-58&t=4705s
I'm unable to understand what Columbus Rites Theater stuff is, with his accent.

Maybe, `Golomb-Rice filter stuff` ? Wasabi Wallet relies on these Golomb-Rice filters? BIP157, 158

And the other issue is that when you have tens of thousands of addresses, like it is taking long to make those tens of thousands of cryptography key challenge exchanges to figure out the entire transaction history.
So this might be taking like half an hour or how long?

> **Audience 4:** 01:14:43

Reviewer: ```suggestion
**Peter:** 01:14:43
```

Reviewer: <img width="1694" height="643" alt="Image" src="https://github.com/user-attachments/assets/2239aa98-8e2c-4eba-aa9d-7ea2035f5692" />

If I may add something, you can have multiple servers, let's say in Amazon, and running them in parallel with Spiral on.
So it should be costly, but still more fast.

**Max Hillebrand:** 01:15:01

Yeah, for your info, this is a CPU heavier, but it's parallelizable.
So if we get a server with 50,000 CPU cores, then it would be super quick.

> **Audience 4:** 01:15:12

Reviewer: ```suggestion
**Peter:** 01:15:12
```

Reviewer: <img width="1704" height="564" alt="Image" src="https://github.com/user-attachments/assets/b1b3b406-71a9-4a11-ac1f-ad9109245adc" />

So, sometimes you want to say, okay, how can I ask for this to a central server without revealing my identity or my intentions, right?
So that kind of things is something that, that's why I say that we have to keep an eye on this technology.

**Samir Menon:** 01:24:33

Yeah, so that's a great point.
So I know I'm kind of more familiar, I have been in cryptography, but not crypto.
So I'm kind of like, I was really into Bitcoin in like 2015, I TA'd our class and then I like totally checked out.
Yeah, so I'm a little unfamiliar, but what kind of metadata, what kind of like data do block explorers provide about an address that is kind of that global state.
So I can think of like in the kind of Ethereum version, I can think of like, some NFT thing, they like, pull up some image there's like that kind of data.
But what are you thinking of that is global that these explorers would show?

**Lucas Ontivero:** 01:25:28

Well, of course we don't know, right?

> Speaker 4: 01:25:31

Reviewer: ?

We were in like a privacy enhancing technologies kind of venue, but it was still, yeah.
I mean, it was a big deal for the reason you pointed out, which is it hasn't been practical.
And I think our work was kind of one of several works that took it from being impractical to practical.
There are now a couple, but at the time we were kind of one of the early ones that, yeah, I mean, when we told people, you go to this site, and we could read Wikipedia, that was like a big deal, because I think everything else had been kind of speculative.
It was like, if you were looking up a sensitive medical condition in theory and if you wanted to you know there's a lot of like hypotheticals.

**Max Hillebrand:** 01:30:39

yeah the MVP demos are pretty great

**Samir Menon:** 01:30:44

yeah if you have ideas for other ones let me know I want to do Ethereum and I want to do, sorry, a password checker.

## Password Checker

**Samir Menon:** 01:30:54

So I know, I don't know if you guys have seen, have I been pwned, but they have like a site where you put in a password to see if it's been leaked but you know of course you're just giving them your password.

> They do some hashing and stuff they say it's okay but you know you could do much better with PiR.

Reviewer: ```suggestion
They do some hashing and stuff they say it's okay but you know you could do much better with PIR.
```

So, that is something you can use homomorphic encryption for.
And then that is amazing, because obviously, who you talk to is really sensitive.
And if we could hide that data, even more than what you say, but also who you who you talk to, that would be really cool.
So that's, something I see as as really realistic.
And we have to do this foundational work of making PIR practical first.
But once we do that, I think building a messenger, I hope I can in five years tell you, yeah, you can go and message people and the service will never learn who you talk to.

**Max Hillebrand:** 01:34:52

It's pretty sweet.
Just a fun fact.
But the Signal groups use the key to verified anonymous credential cryptography scheme that we use for our Wasabi CoinJoins.
So it would be nice if we throw your homomorphic encryption on top.

**Samir Menon:** 01:35:10

So what is that called?
It's a key, what is key?

**Max Hillebrand:** 01:35:14

Key verified anonymous credentials.

**Samir Menon:** 01:35:19

Okay.

**Max Hillebrand:** 01:35:20

This is for context or for a setup of the conversation?

> **Samir Menon:** 01:35:21

Reviewer 2: No words below Samir Menon

Author: He wasn't a active speaker during this timestamp

So, that is something you can use homomorphic encryption for.
And then that is amazing, because obviously, who you talk to is really sensitive.
And if we could hide that data, even more than what you say, but also who you who you talk to, that would be really cool.
So that's, something I see as as really realistic.
And we have to do this foundational work of making PIR practical first.
But once we do that, I think building a messenger, I hope I can in five years tell you, yeah, you can go and message people and the service will never learn who you talk to.

**Max Hillebrand:** 01:34:52

It's pretty sweet.
Just a fun fact.
But the Signal groups use the key to verified anonymous credential cryptography scheme that we use for our Wasabi CoinJoins.
So it would be nice if we throw your homomorphic encryption on top.

**Samir Menon:** 01:35:10

So what is that called?
It's a key, what is key?

**Max Hillebrand:** 01:35:14

> Key verified anonymous credentials.

Reviewer: ```suggestion
Key verified anonymous credentials.  (keyed-verification anonymous credentials, or KVAC)
```

Author: I don't think we need this here

So, that is something you can use homomorphic encryption for.
And then that is amazing, because obviously, who you talk to is really sensitive.
And if we could hide that data, even more than what you say, but also who you who you talk to, that would be really cool.
So that's, something I see as as really realistic.
And we have to do this foundational work of making PIR practical first.
But once we do that, I think building a messenger, I hope I can in five years tell you, yeah, you can go and message people and the service will never learn who you talk to.

**Max Hillebrand:** 01:34:52

It's pretty sweet.
Just a fun fact.
But the Signal groups use the key to verified anonymous credential cryptography scheme that we use for our Wasabi CoinJoins.
So it would be nice if we throw your homomorphic encryption on top.

**Samir Menon:** 01:35:10

So what is that called?
It's a key, what is key?

**Max Hillebrand:** 01:35:14

> Key verified anonymous credentials.

Reviewer: ```suggestion
Key verified anonymous credentials. (Keyed-Verification Anonymous Credentials)
```

Reviewer: https://github.com/bitcointranscripts/bitcointranscripts/pull/643/changes#r3162343619

@0tuedon The speaker uses `Key verified anonymous credentials` incorrectly in three places. As discussed, we need the correct terms for the "Knowledge base." What do you think about just appending the first one?

Reviewer 2: Like in this same pattern yh

Author: So we make changes to all or just first one ?

Reviewer: I would just VScode Find and Replace all. lol

@0tuedon What would you prefer? It's your call

Author: Yeah I wasn't sure to apply to all

**Samir Menon:** 01:35:19

Okay.

**Max Hillebrand:** 01:35:20

This is for context or for a setup of the conversation?

**Max Hillebrand:** 01:35:23

We use it as basically eCash token for access rights.

**Samir Menon:** 01:35:29

Okay.

**Max Hillebrand:** 01:35:30

The token gets created during input registration and you can only register an output if you present such a token and it's anonymous e-cash so to say.
Very rough explanation.

**Adam Ficsor:** 01:35:43

Cool.

> Are you familiar with the chaumian blind signatures?

Reviewer: ```suggestion
Are you familiar with the Chaumian blind signatures?
```

Reviewer: From the previous review:
https://github.com/bitcointranscripts/bitcointranscripts/pull/643#discussion_r3162387805

**Samir Menon:** 01:35:49

> chaumian blind signatures?

Reviewer: ```suggestion
Chaumian blind signatures?
```

Yes.

> I'm familiar with the name chaumian, but no, I haven't heard of chaumian blind signatures.

Reviewer: ```suggestion
I'm familiar with the name Chaumian, but no, I haven't heard of Chaumian blind signatures.
```

But it's something I mean, I don't know exactly how to define homomorphic encryption, right?

## Homomorphic Encryption

**Lucas Ontivero:** 01:37:51

Because homomorphic encryption, like, in the concept, It's okay, but it sounds like something that is not still possible, right?
So homomorphic encryption, and the scope is this for most of the time.
It's something like what we do is basically we operate in a committed value and that value is basically a point in an elliptic curve and that's it.

**Samir Menon:** 01:38:36

It's funny, I think elliptic curves are way more complicated.
I'm always like, I think lattice-based cryptography and homomorphic stuff, which is lattice-based and was initially pioneered for post-quantum resistance and stuff.
I think it's way easier to understand than elliptic curves.
Because what is an elliptic curve, really?
I'm always very confused.
Not confused, but it's always very abstract for me.
Whereas lattices it's matrices and it's noise.
And I understand that a lot better.
It's like a lot easier for me.
So it's funny, you should you guys are real crypto people.

> If you guys are just sitting around thinking about Pedersen commitments and homo.

Reviewer: ```suggestion
If you guys are just sitting around thinking about Pedersen commitments and homomorphisms.
```

> If you guys are just sitting around thinking about Henderson commitments and homo.

Reviewer: Youtube captions
<img width="1962" height="374" alt="Image" src="https://github.com/user-attachments/assets/ca814364-f84c-4f50-b16d-6e0b46bc8bd0" />

> If you guys are just sitting around thinking about Henderson commitments and homo.

Reviewer: ```suggestion
If you guys are just sitting around thinking about Pedersen commitments and homo.
```

There's kind of like, that's a very specific homomorphism for ECC, but yeah.
Yeah, they're connected.

**Max Hillebrand:** 01:37:06

> As far as I remember we do use peterson commitments or generally speaking homomorphic encryption for the amount of the value of these credentials.

Reviewer: ```suggestion
As far as I remember we do use Pedersen commitments or generally speaking homomorphic encryption for the amount of the value of these credentials.
```

There's kind of like, that's a very specific homomorphism for ECC, but yeah.
Yeah, they're connected.

**Max Hillebrand:** 01:37:06

> As far as I remember we do use pedersen commitments or generally speaking homomorphic encryption for the amount of the value of these credentials.

Reviewer: ```suggestion
As far as I remember we do use Pedersen commitments or generally speaking homomorphic encryption for the amount of the value of these credentials.
```

Reviewer: From the previous review:
https://github.com/bitcointranscripts/bitcointranscripts/pull/643#discussion_r3162431540

**Samir Menon:** 01:37:15

Ah okay

> **Audience 4:** 01:37:17

Reviewer: ```suggestion
**Lucas Ontivero:** 01:37:17
```

Reviewer: <img width="2278" height="1099" alt="Image" src="https://github.com/user-attachments/assets/bb51933f-8aa1-4bfe-9e30-a1f256214c2e" />

But it does exist.
It is possible.

**Audience 2:** 01:45:15

I don't suppose you can make a mini computer out of a fully homomorphic interface.

**Samir Menon:** 01:45:22

No, you absolutely can.
In fact, people are working on it for roll-ups and stuff.
Yeah, it's possible.

**Audience 2:** 01:45:30

If I may ask Fabricio, is that implementable in lookup tables and FPGAs and stuff?

**Samir Menon:** 01:45:40

Yeah, it's not that amenable to hardware acceleration.

**Audience 3:** 01:45:44

You preceded me with the question I was going to ask.

> If the implementation on FBAs or ASICs could constitute a significant improvement.

Reviewer: ```suggestion
If the implementation on FPGAs or ASICs could constitute a significant improvement.
```

But it does exist.
It is possible.

**Audience 2:** 01:45:15

I don't suppose you can make a mini computer out of a fully homomorphic interface.

**Samir Menon:** 01:45:22

No, you absolutely can.
In fact, people are working on it for roll-ups and stuff.
Yeah, it's possible.

**Audience 2:** 01:45:30

If I may ask Fabricio, is that implementable in lookup tables and FPGAs and stuff?

**Samir Menon:** 01:45:40

Yeah, it's not that amenable to hardware acceleration.

> Speaker 7: 01:45:44

Reviewer 2: Speaker 7?

> **Audience 3:** 01:45:44

Reviewer: ```suggestion
**Fabrizio Armani:** 01:45:44
```

Reviewer: <img width="980" height="859" alt="Image" src="https://github.com/user-attachments/assets/b669899f-76b9-412c-b767-7f4914b9c392" />

Reviewer: https://black-paper.rgb.tech/other-information/13.-history-and-acknowledgements
https://voskuil.org/cryptoeconomics/cryptoeconomics.pdf

And it was it was fast, but it wasn't very cost effective.
Now, now that GPUs are way cheaper, it might be.
I have to have to like revisit it.

**Max Hillebrand:** 01:47:05

I show you a question because you just said the database has to be kept in memory.
Like, is that the case in Spiral?

**Samir Menon:** 01:47:11

Yes.

**Max Hillebrand:** 01:47:12

Oh, okay.
But then, I mean, if we have a large data data set, for example, the full blockchain, many hundreds of gigabytes.
Right.
All of that needs to be in memory?

**Samir Menon:** 01:47:23

Yeah, yeah, it does.
I will say the memory cost usually kind of pales in terms of in comparison to the compute cost.
So I mean, making everything fit in memory is, I mean, hundreds of gigs of memory is on a cloud provider, not that expensive, actually.
And once you have everything fitting in memory, you can throw as many compute cores as you want at it.
So it's a limit, but it's not crazy.

**Audience 2:** 01:47:55

Out of topic question for Fabricio.
I've been quite interested in FPGA development.
What's this?

> Have you looked into the open source hardware's verilog synthesis stuff like Ice Storm or something?

Reviewer: ```suggestion
Have you looked into the open source hardware's Verilog synthesis stuff like IceStorm or something?
```

https://ocw.mit.edu/courses/6-884-complex-digital-systems-spring-2005/pages/related-resources/

Reviewer: https://github.com/YosysHQ/icestorm
Project `IceStorm`

Now, now that GPUs are way cheaper, it might be.
I have to have to like revisit it.

**Max Hillebrand:** 01:47:05

I show you a question because you just said the database has to be kept in memory.
Like, is that the case in Spiral?

**Samir Menon:** 01:47:11

Yes.

**Max Hillebrand:** 01:47:12

Oh, okay.
But then, I mean, if we have a large data data set, for example, the full blockchain, many hundreds of gigabytes.
Right.
All of that needs to be in memory?

**Samir Menon:** 01:47:23

Yeah, yeah, it does.
I will say the memory cost usually kind of pales in terms of in comparison to the compute cost.
So I mean, making everything fit in memory is, I mean, hundreds of gigs of memory is on a cloud provider, not that expensive, actually.
And once you have everything fitting in memory, you can throw as many compute cores as you want at it.
So it's a limit, but it's not crazy.

**Jumar Macato:** 01:47:55

Out of topic question for Fabricio.
I've been quite interested in FPGA development.
What's this?
Have you looked into the open source hardware's Verilog synthesis stuff like IceStorm or something?
Ah, sorry, project starts with that.
I forgot.

> **Audience 3:** 01:48:23

Reviewer: ```suggestion
**Fabrizio Armani:** 01:48:23
```

Reviewer: <img width="1032" height="824" alt="Image" src="https://github.com/user-attachments/assets/427bad9d-3c22-4ac8-a161-075f5baaf7c2" />

But actually, these are crypto project guarding FPGA, because actually I'm dealing with FPGA for actually for genomic recognition.
But actually regarding crypto, I actually I didn't look yet, let's say.
But actually if you, let's say regarding, for example, the primitive function to, for example, do hashing stuff or cryptographic primitives, I think you can improve probably one order of magnitude if you implement some cryptographic computation and hashing on FPGAs. This is something that I think is worth looking at, definitely.

**Jumar Macato:** 01:49:28

Cool.
Last question for you.
Has it been feasible, given that is it been in FPGAs, you are bound with a manufacturer's tooling?
And lately, there has been some developments regarding open source tooling for like Yosys and stuff.
Have you used those tools before or is it still vendor locked?

> **Audience 3:** 01:50:03

Reviewer: ```suggestion
**Fabrizio Armani:** 01:50:03
```

Reviewer: <img width="1013" height="766" alt="Image" src="https://github.com/user-attachments/assets/ddeb5cc4-4192-4960-b116-0b79a5374969" />

> Well at the moment we are using proper proprietary tools, for example Vivado, which is very famous, but actually, no, actually didn't yet look at the say open source opportune source development regarding that I will have a look

Reviewer: ```suggestion
Well at the moment we are using proper proprietary tools, for example Vivado, which is very famous, but actually, no, actually didn't yet look at the say open source development regarding that I will have a look
```

Reviewer: @Garvit-77 Please delete the extra typo.

@0tuedon Ready to merge

**Jumar Macato:** 01:50:33

so much thank you so much

**Max Hillebrand:** 01:50:37

well Samir, thank you very much for coming here.
That was really kind of you.

**Samir Menon:** 01:50:43

Yeah, thanks so much.
I learned so much.
I'm learning so much about what Bitcoin people want, the need and stuff.
You guys are super helpful.
So thank you so much.
I want to, the only thing I want to say is, let's keep in touch, especially as we get closer to building something that looks less like just a balance checker and more like a private block explorer, I would love to see if there's kind of a synergy between what we offer.
I would love to see if people who use Wasabi Wallet also wanna use a private block Explorer.
So yeah, if that sounds good, I would love to keep in touch.

**Max Hillebrand:** 01:51:21

Yeah, definitely.

> I mean, especially if you run the server and we don't have to, and we can use it for small niche things, like finding out the amount of the input of a payment transaction so that we can pre-bump it.

Reviewer: ```suggestion
I mean, especially if you run the server and we don't have to, and we can use it for small niche things, like finding out the amount of the input of a payment transaction so that we can fee-bump it.
```

Ah, sorry, project starts with that.
I forgot.

**Audience 3:** 01:48:23

But actually, these are crypto project guarding FPGA, because actually I'm dealing with FPGA for actually for genomic recognition.
But actually regarding crypto, I actually I didn't look yet, let's say.
But actually if you, let's say regarding, for example, the primitive function to, for example, do hashing stuff or cryptographic primitives, I think you can improve probably one order of magnitude if you implement some cryptographic computation and hashing on FPGAs. This is something that I think is worth looking at, definitely.

**Audience 2:** 01:49:28

Cool.
Last question for you.
Has it been feasible, given that is it been in FPGAs, you are bound with a manufacturer's tooling?

> And lately, there has been some developments regarding open source tooling for like uses and stuff.

Reviewer: ```suggestion
And lately, there has been some developments regarding open source tooling for like Yosys and stuff.
```
https://github.com/yosyshq/yosys
